src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py

```diff
@@ -670,7 +669,11 @@ def build_binary(
             # duplicated queries if there is a common CTE block referenced by
             # both left and right.
             referenced_ctes.update(select_left.referenced_ctes)
-            referenced_ctes.update(select_right.referenced_ctes)
+            for with_query_block, count in select_right.referenced_ctes.items():
```

Comment:
I see a lot of if else, i think you can add a utility function in the query_plan_analysis_utils.py to help merge the refered_ctes which takes two dict.